pkg/ebpf/events_pipeline.go

```diff
@@ -281,35 +282,47 @@ func (t *Tracee) decodeEvents(outerCtx context.Context, sourceChan chan []byte)
 	return out, errc
 }
 
-// computePolicies iterates through the policies that do the filtering in user space, checking whether an event should be considered.
-// If it should not, it sets the respective offset to 0.
-// Finally it returns computed policies and the names of the policies that matched.
-func (t *Tracee) computePolicies(event *trace.Event) (uint64, []string) {
+// computePoliciesMatches iterates through the policies that do the filtering in user space.
+// If the filter doesn't match, the policy bit is cleared from the matched policies bitfield.
+// Finally, returns the bitmap with filtered matches (to be propagated in the pipeline) and
+// the names of the policies that matched (actual matches), without making any changes to the event.
+func (t *Tracee) computePoliciesMatches(event *trace.Event) (bitmap uint64, names []string) {
 	eventID := events.ID(event.EventID)
-	matchedPolicies := event.MatchedPolicies
-	matchedPoliciesNames := []string{}
+	bitmap = event.MatchedPolicies
 
 	for p := range t.config.Policies.Map() {
 		bitOffset := uint(p.ID)
 
 		// Events submitted with matching policies.
 		// The policy must have its bit cleared when it does not match.
-		if !utils.HasBit(matchedPolicies, bitOffset) {
+		if !utils.HasBit(bitmap, bitOffset) {
+			continue
+		}
+
+		// An event matched in a policy might not match all policies.
+		_, ok := p.EventsToTrace[eventID]
+		if !ok {
+			// The remaining filter checks can be skipped, allowing to continue to the next policy.
+			// This treats the current policy as unmatched, even though this event could potentially
+			// match as a derived event in later stages. That is why it is crucial to propagate
+			// the event matching bit as-is, enabling checks for possible derived events.
+
+			// Therefore, the field to be used for actual policy matching checks is the policy name.
 			continue
 		}
```

Comment:
In the sink stage we do the following:
```
// Only emit events requested by the user
id := events.ID(event.EventID)
event.MatchedPolicies &= t.events[id].emit
if event.MatchedPolicies == 0 {
    continue
}
```

1. If we would have done the mapping to policy names only at the sink stage (after the lines I pasted above) I believe this would have also solve the problem, no?
2. Let's assume we only requested for a derived event in some policy. So now the policy bitmap (MatchedPolicies) is not zero, but the matchedPoliciesNames is empty for the original event, right? Won't this cause the original event, which should not be printed, to be emitted (with an empty matched policy list)?

Reply:
> 2\. Let's assume we only requested for a derived event in some policy. So now the policy bitmap (MatchedPolicies) is not zero, but the matchedPoliciesNames is empty for the original event, right? Won't this cause the original event, which should not be printed, to be emitted (with an empty matched policy list)?

Actually the lines I pasted above should resolve this I believe.

And about the first point I wrote - is there a reason not to do the bitmap to names mapping in the sink stage (even if only for performance reasons)

Reply:
I think that eventually we should do the following:
1. Keep the original matched policies bitmap we got from the kernel untouched, to be used by derived events as well
2. Make a copy of this bitmap, which will be used to check for matched policies of the current events
3. Continue with the pipeline in any case if the event has derived events (like we do now)
4. In the sink stage, only emit the event if the copy of the bitmap is non zero and event should be emitted
5. Map the matchedPolicies bitmap to policy name in the sink stage if we passed all above

WDYT?

Reply:
@yanivagman, I'm doing some tests and what you proposed makes perfect sense:

![image](https://github.com/aquasecurity/tracee/assets/3372117/c265a805-c1bb-4562-a650-a7b56b5a8622)

I was actually going to the same direction after @rafaeldtinoco's findings in https://github.com/aquasecurity/tracee/issues/3184#issuecomment-1571725177:

>  I'm really considering that we have to propagate the MatchedScopes bitmap without bit clearing at all.

@josedonizetti and @NDStrahilevitz FYI